spec/unit/puppetfile_spec.rb

```diff
@@ -302,12 +283,12 @@ def expect_wrapped_error(orig, pf_path, wrapped_error)
         block.call
       end
 
-      mod1 = spy('module')
+      mod1 = double('module', :cachedir => :none)
```

Comment:
Forgive me for not knowing much about `spy`s, but why did all the `spy`s change to `double`s?

Reply:
> Spies are [a] type of test double that support [...] allowing you to expect that a message has been received after the fact, using have_received.

Spies are the least specific kind of double, and allow any method to be called on them. However, no matter what method is called, a spy pretty much always just returns itself.

For this test, I'd like the doubles to respond to a specific method, `#cachedir`, with a common value. This is because in [`32af983`](https://github.com/puppetlabs/r10k/pull/1120/commits/32af9836ae1fbfa65df1b76bb06b2df44f198dcf#diff-1e18cef8c5fe2a6b1b2146ea4282b5bb0e40cf5f6fec5bdae9112d0830cf12fcR217) we implemented grouping of modules using `#group_by { |mod| mod.cachedir }`. So, I upgraded the spies to the next level of specificity, doubles, so that they will report a known value when interrogated for their `#cachedir`.

> When creating a double, you can allow messages (and set their return values) by passing a hash.

Thinking about it now I'm not really testing for how the queue has absorbed or grouped these modules so it kinda doesn't matter, I guess. However, I think in the commit to improve variable names I'll upgrade them to verifying doubles, which will at least have them alert if any methods are called on them that aren't valid module methods. As a non-spy verifying double it becomes required to let them expect that they're gonna have `#cachedir` called on them, and so makes sense to give them a value to return for it.

Reply:
gotcha, thanks so much for this explanation!